File: day7.py
with open("day7.txt") as reader:
    content = reader.read().strip().split('\n')

content = [c.split(":") for c in content]
results = [int(r[0]) for r in content]
operands = [tuple(map(int,r[1].strip().split(" "))) for r in content]


def to_base3(n):
    if n==0: return "0"
    res = ""
    while n!=0:
        res = str(n%3) + res
        n = n//3
    return res

def op(x,y,op):
    if op=="2":
        return int(str(x)+str(y))
    return x+y if op=="1" else x*y

def compute(operands, operations):
    assert len(operations) == len(operands) -1
    res = op(operands[0],operands[1],operations[0])
    for i in range(1,len(operations)):
        res = op(res, operands[i+1],operations[i])
    return res

def check(result, operands):
    operations = 0
    size = len(operands)-1
    while operations<=2**(size)-1:
        to_do = bin(operations)[2:].zfill(size)
        if compute(operands,to_do)==result:
            return True,result
        operations +=1
    return False, 0


def check3(result, operands):
    operations = 0
    size = len(operands)-1
    while operations<3**size:
        to_do = str(to_base3(operations)).zfill(size)
        if compute(operands,to_do)==result:
            return True,result
        operations +=1
    return False, 0

values = [check(res, oper) for (res,oper) in zip(results, operands)]
p1 = sum([v[1] for v in values])
print(f"Résultat partie 1 = {p1}")

values = [check3(res, oper) for (res,oper) in zip(results, operands)]
p2 = sum([v[1] for v in values])
print(f"Résultat partie 1 = {p2}")

File: day7.c
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

#define MSIZE 100
#define LINESIZE 500

struct equation
{
    uint64_t result;
    int size;
    uint64_t operands[MSIZE];
};
typedef struct equation equation;

equation read(FILE *reader)
{
    equation c;
    if (fscanf(reader, "%ld: ", &c.result)!=1)
        {c.size=0;
        return c;}
    char buffer[LINESIZE];
    fgets(buffer, sizeof(buffer), reader);
    char *op = strtok(buffer, " ");
    int i = 0;
    while (op != NULL)
    {
        c.operands[i++] = atoi(op);
        op = strtok(NULL, " ");
    }
    c.size = i;
    return c;
}

void display(equation c)
{
    for (int i = 0; i < c.size; i++)
    {
        printf("%ld; ", c.operands[i]);
    }
    printf(" -> %ld\n", c.result);
}

bool check(equation c, int cop, uint64_t target)
{
    if (cop == 1)
        return ((target == c.operands[0] + c.operands[1] || target == c.operands[0] * c.operands[1]));
    bool addok = (target >= c.operands[cop]);
    bool multok = (target % c.operands[cop] == 0);
    return (addok && check(c, cop - 1, target - c.operands[cop])) || (multok && check(c, cop - 1, target / c.operands[cop]));
}

int main()
{
    FILE *reader = fopen("day7.txt", "r");
    equation c;
    uint64_t res = 0;
    while (!feof(reader))
    {
        c = read(reader);
        if (c.size!=0 && check(c, c.size - 1, c.result))
        {
            res += c.result;
        }
    }
    printf("Réponse partie 1 = %ld\n", res);
}